src/random.cpp

```diff
@@ -78,25 +78,84 @@ static inline int64_t GetPerformanceCounter() noexcept
 }
 
 #if defined(__x86_64__) || defined(__amd64__) || defined(__i386__)
-static bool rdrand_supported = false;
+static bool g_rdrand_supported = false;
+static bool g_rdseed_supported = false;
 static constexpr uint32_t CPUID_F1_ECX_RDRAND = 0x40000000;
+static constexpr uint32_t CPUID_F7_EBX_RDSEED = 0x00040000;
+
+static void inline GetCPUID(uint32_t leaf, uint32_t subleaf, uint32_t& a, uint32_t& b, uint32_t& c, uint32_t& d)
+{
+#ifdef __GNUC__
+    __cpuid_count(leaf, subleaf, a, b, c, d);
+#else
+  __asm__ ("cpuid" : "=a"(a), "=b"(b), "=c"(c), "=d"(d) : "0"(leaf), "2"(subleaf));
+#endif
+}
+
 static void InitHardwareRand()
 {
     uint32_t eax, ebx, ecx, edx;
-    if (__get_cpuid(1, &eax, &ebx, &ecx, &edx) && (ecx & CPUID_F1_ECX_RDRAND)) {
-        rdrand_supported = true;
+    GetCPUID(1, 0, eax, ebx, ecx, edx);
+    if (ecx & CPUID_F1_ECX_RDRAND) {
+        g_rdrand_supported = true;
+        GetCPUID(7, 0, eax, ebx, ecx, edx);
+        if (ebx & CPUID_F7_EBX_RDSEED) {
+            g_rdseed_supported = true;
+        }
     }
 }
 
 static void ReportHardwareRand()
 {
-    if (rdrand_supported) {
-        // This must be done in a separate function, as HWRandInit() may be indirectly called
-        // from global constructors, before logging is initialized.
+    // This must be done in a separate function, as HWRandInit() may be indirectly called
+    // from global constructors, before logging is initialized.
+    if (g_rdseed_supported) {
+        LogPrintf("Using RdSeed and RdRand as additional entropy sources\n");
+    } else if (g_rdrand_supported) {
         LogPrintf("Using RdRand as an additional entropy source\n");
     }
 }
 
+/** Read 64 bits of entropy using rdrand.
+ *
+ * Must only be called when RdRand is supported.
+ */
+static uint64_t GetRdRand() noexcept
+{
+    uint8_t ok;
+#ifdef __i386__
+    uint32_t r1, r2;
+    __asm__ volatile (".byte 0x0f, 0xc7, 0xf0, 0x0f, 0xc7, 0xf2; setc %2" : "=a"(r1), "=d"(r2), "=q"(ok) :: "cc"); // rdrand %eax; rdrand %edx
+    assert(ok);
+    return (((uint64_t)r2) << 32) | r1;
+#else
```

Comment:
I feel like this and the one below should be guarded by #if defined(__x86_64__) || defined(__amd64__)  ... I think we'd like to prefer x86_64 asm in the output, even if we're not going to call it. :P